File: src/main/java/com/planbase/pdf/layoutmanager/XyDim.java
// Copyright 2014-08-18 PlanBase Inc. & Glen Peterson
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.planbase.pdf.layoutmanager;

/**
 Represents a 2D dimension in terms of X and Y where both X and Y must be non-negative.
 Do not confuse a dimension (measurement) with an XyOffset which represents coordinates where
 negative y is down from the upper left-hand corner.
 */
public class XyDim {
    public static final XyDim ZERO = new XyDim(0, 0) {
        @Override public int hashCode() { return 0; }
    };
    private final double x;
    private final double y;
    private XyDim(double xCoord, double yCoord) {
        if ( (xCoord < 0) || (yCoord < 0) ) {
            throw new IllegalArgumentException("Dimensions must be positive");
        }
        x = xCoord; y = yCoord;
    }
    public static XyDim of(double x, double y) {
        if ((x == 0) && (y == 0)) { return ZERO; }
        return new XyDim(x, y);
    }
    public double x() { return x; }
    public double y() { return y; }
    public XyDim x(double newX) { return of(newX, y); }
    public XyDim y(double newY) { return of(x, newY); }

    public XyDim minus(XyDim that) { return of(this.x - that.x(), this.y - that.y()); }
    public XyDim plus(XyDim that) { return of(this.x + that.x(), this.y + that.y()); }

//    public XyPair plusXMinusY(XyPair that) { return of(this.x + that.x(), this.y - that.y()); }

    public XyDim maxXandY(XyDim that) {
        if ((this.x >= that.x()) && (this.y >= that.y())) { return this; }
        if ((this.x <= that.x()) && (this.y <= that.y())) { return that; }
        return of((this.x > that.x()) ? this.x : that.x(),
                  (this.y > that.y()) ? this.y : that.y());
    }
//    public XyPair maxXMinY(XyPair that) {
//        if ((this.x >= that.x()) && (this.y <= that.y())) { return this; }
//        if ((this.x <= that.x()) && (this.y >= that.y())) { return that; }
//        return of((this.x > that.x()) ? this.x : that.x(),
//                  (this.y < that.y()) ? this.y : that.y());
//    }

    /** Compares dimensions */
    public boolean lte(XyDim that) { return (this.x <= that.x()) && (this.y <= that.y()); }

    @Override public String toString() { return "XyDim(" + x + " " + y + ")"; }

    @Override public int hashCode() { return Double.hashCode(x) ^ Double.hashCode(y); }

    @Override public boolean equals(Object other) {
        // Cheapest operation first...
        if (this == other) { return true; }

        if ( (other == null) ||
             !(other instanceof XyDim) ||
             (this.hashCode() != other.hashCode()) ) {
            return false;
        }
        // Details...
        final XyDim that = (XyDim) other;

        // Compare "significant" fields here.
        return (x == that.x()) && (y == that.y());
    }
}

File: src/main/java/com/planbase/pdf/layoutmanager/XyDim.kt
// Copyright 2017 PlanBase Inc.
//
// This file is part of PdfLayoutMgr
//
// PdfLayoutMgr is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// PdfLayoutMgr is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with PdfLayoutMgr.  If not, see <https://www.gnu.org/licenses/agpl-3.0.en.html>.
//
// If you wish to use this code with proprietary software,
// contact PlanBase Inc. <https://planbase.com> to purchase a commercial license.

package com.planbase.pdf.layoutmanager

import org.apache.pdfbox.pdmodel.common.PDRectangle

/**
 * Immutable 2D dimension in terms of non-negative width and height.
 * Do not confuse a dimension (measurement) with an XyOffset which represents coordinates where
 * the bottom of the page is zero and positive height is up from there.
 * Remember: an XyDim on a Portrait orientation page has the width and height *reversed*.
 */
data class XyDim(val width: Float, val height: Float) {
    init {
        if (width < 0 || height < 0) {
            throw IllegalArgumentException("Dimensions must be positive")
        }
    }

    /**
     * Returns an XyDim with the width and height taken from the same-named fields on the given
     * rectangle.
     */
    constructor(rect: PDRectangle) : this(rect.width, rect.height)


    fun width(newX: Float) = XyDim(newX, height)

    fun height(newY: Float) = XyDim(width, newY)

    /**
     * If true, returns this, if false, returns a new XyDim with width and height values swapped.
     * PDFs think the long dimension is always the height of the page, regardless of portrait vs.
     * landscape, so we need to conditionally adjust what we call width and height.
     */
    // This is suspicious because we're swapping width and height and the compiler thinks
    // we might be doing so by accident.
    fun swapWh() = XyDim(height, width)

    /** Returns a PDRectangle with the given width and height (but no/0 offset)  */
    fun toRect() = PDRectangle(width, height)

    //    /** Returns a PDRectangle with the given width and height (but no/0 offset) */
    //    public PDRectangle toRect(XyOffset off) {
    //        return new PDRectangle(off.x(), off.y(), width, height);
    //    }

    /** Subtracts the given XyDim from this one (remember, these can't be negative).  */
    operator fun minus(that: XyDim) = XyDim(this.width - that.width, this.height - that.height)

    /** Adds the given XyDim from this one  */
    operator fun plus(that: XyDim) = XyDim(this.width + that.width, this.height + that.height)

    //    public XyPair plusXMinusY(XyPair that) { return of(this.width + that.width(), this.height - that.height()); }

    //    public XyDim maxXandY(XyDim that) {
    //        if ((this.width >= that.width()) && (this.height >= that.height())) { return this; }
    //        if ((this.width <= that.width()) && (this.height <= that.height())) { return that; }
    //        return of((this.width > that.width()) ? this.width : that.width(),
    //                  (this.height > that.height()) ? this.height : that.height());
    //    }

    /** Compares dimensions and returns true if that dimension doesn't extend beyond this one.  */
    fun lte(that: XyDim): Boolean = this.width <= that.width &&
                                    this.height <= that.height

    companion object {
        val ZERO = XyDim(0f, 0f)
    }
}
